keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,423 @@
+     updates the Container resource limit, and then updates Pod-level cgroup limit.
+  1. If resource update results in no net change of a resource type, only the Container
+     resource limits are updated.
+In all the above cases, Kubelet applies Container resource limit decreases before
+applying limit increases.
+
+#### Container resource limit update failure handling
+
+For simplicity, if Container resource limits update fails, Kubelet restarts the
+Container in-place to allow new limits to take effect, and the action is logged.
+
+#### Notes
+
+* To avoid races and possible gamification, all components should use
+  max(ResourceRequirements, ResourcesAllocated) when computing resources used by
+  a Pod. TBD if this can be weakened when PodResizeSuccess PodCondition's
+  Reason is set to Rejected, or should the initiating actor update
+  ResourceRequirements back to reclaim resources.
+* If another resource update arrives when a previous update is being handled,
+  that and all subsequent updates should be buffered at the Controller, and
+  applied upon success/failed completion of the update that is in progress.
+* Lowering memory limits may not always work if the application is holding on
+  to pages. Kubelet will use a control loop to set the memory limits near usage
+  in order to force a reclaim, and update ResourcesAllocated only when limit is
+  at desired value.
+* Impact of Pod Overhead: Scheduler adds Pod Overhead to the resize request
+  to determine if in-place resize is possible. Kubelet implements Pod Overhead
+  if the values have changed.
+* Impact of memory-backed emptyDir volumes: If memory-backed emptyDir is in use,
+  Kubelet will clear out any files in emptyDir upon Container restart.
+
+### Affected Components
+
+Pod v1 core API:
+* extended model,
+* added validation.
+
+Admission Controllers: LimitRanger, ResourceQuota need to support Pod Updates:
+* for ResourceQuota it should be enough to change podEvaluator.Handler
+  implementation to allow Pod updates; max(ResourceRequirements, ResourcesAllocated)
+  should be used to be in line with current ResourceQuota behavior
+  which blocks resources before they are used (e.g. for Pending Pods),
+* for LimitRanger TBD.
+
+Kubelet:
+* support in-place resource management,
+* set PodStatus ResourcesAllocated for Containers on placing the Pod on Node.
+* change UpdateContainerResources CRI API so that it works for both Linux and Windows.
+
+Scheduler:
+* determine if in-place resize is possible, updates its cache depending on resizing outcome.
+* updates its cache based on resizing action by Kubelet.
+
+Controllers:
+* propagate Template resources update to running Pod instances.
+* initiate resource update retries or reschedule Pod (controlled by policy) that failed resize.
+
+Other components:
+* check how the change of meaning of resource requests influence other kubernetes components.
+
+### Possible Extensions
+
+1. Allow resource limits to be updated too (VPA feature).
```

Comment:
is this proposal only allowing requests to be updated and not limits ?

Reply:
We update both. OTOH, VPA monitored and updated requests, not limits when I last checked, but that was some time ago.